### PR TITLE
Enable f5-agent upload pkcs7 intermediates certificate to Bigip

### LINKED ISSUE #1323: Enable f5-openstack-agent to upload pkcs7 certificate to Bigip
#### OpenStack Release
For all OpenStack releases >= M

#### Description
The f5-openstack-agent cannot upload pkcs7 format certificate to a Bigip device. since the Bigip device does not accept pkcs7 format certificate for now.  

So this new feature enables f5-openstack-agent to converts pkcs7 format certificate to a x509 format certificate chain and upload the chain to the Bigip device as a bundle. 

The pkcs7 certificate is retrieved from the Openstack Barbican secret container and converted to x509 certificate chain by f5-openstack-agent.

#### Agent Version
version 9.6

### PR DESCRIPTION
**Pull request by zhang-shengping**
Problem: f5-oslbaasv2-agent cannot upload the pkcs7 certificate to Bigip,
since it has no function to do pkcs7 to x509 conversion

Add cert_parser.py file to parse pkcs7 certificate and returns an x509 python
generator, then use the generator produces x509 certificates chain for
ssl_client profile configuration.

@kaustriaf5 
@zhaoqin-github

#### What issues does this address?
Fixes #1323 

#### What's this change do?
Enable f5-agent upload pkcs7 intermediates certifcate to Bigip

#### Where should the reviewer start?
start from create_listener method in f5_openstack_agent/lbaasv2/drivers/bigip/listener_service.py file

#### Any background context?
Some users may need this function to upload the pkcs7 certificates.

### CODE REVIEW

**Comment by zhaoqin-github** [comment]
@zhang-shengping I notice a spelling mistake in your commit message. :-)

File: f5_openstack_agent/lbaasv2/drivers/bigip/test/test_ssl_profile.py [hunk 1]
```diff
@@ -117,3 +120,41 @@ def test_create_client_ssl_profile_raises(self):
...
+            partition='Common',
+            certKeyChain=[
+                {'name': 'testprofile',
+                 'cert': '/Common/testprofile.crt',
+                 'chain': '/Common/testprofile_inter.crt',
+                 'key': '/Common/testprofile.key'}
+            ],
+            sniDefault=False,
+            defaultsFrom=None,
+        )
```
Comment (kaustriaf5) [inline f5_openstack_agent/lbaasv2/drivers/bigip/test/test_ssl_profile.py]: test case needed for scenario with both `parent_profile` and `intermediates`

Comment (kaustriaf5) [reply f5_openstack_agent/lbaasv2/drivers/bigip/test/test_ssl_profile.py]: `test_create_client_ssl_profile_raises` also need to be updated with the new method signature

File: f5_openstack_agent/utils/cert_parser.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    :param certificate_der: Certificate in DER format
+    :returns: crypto high-level x509 data from the DER-encoded certificate
+    """
+    try:
+        x509cert = x509.load_der_x509_certificate(certificate_der,
+                                                  backends.default_backend())
+    except Exception:
+        LOG.exception('Unreadable Certificate.')
+        raise f5_ex.UnreadableCert
+    return x509cert
```
Comment (kaustriaf5) [inline f5_openstack_agent/utils/cert_parser.py]: Do we want to add some basic test coverage for this `cert_parser`?

File: f5_openstack_agent/utils/cert_parser.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+from cryptography.hazmat import backends
+from cryptography.hazmat.primitives import serialization
+from cryptography import x509
+from f5_openstack_agent.lbaasv2.drivers.bigip import exceptions as f5_ex
+from oslo_log import log as logging
+from pyasn1.codec.der import decoder as der_decoder
+from pyasn1.codec.der import encoder as der_encoder
+from pyasn1_modules import rfc2315
+import six
```
Comment (kaustriaf5) [inline f5_openstack_agent/utils/cert_parser.py]: Are these modules included as dependencies by default or expected to be installed as part of the env? Is it the convention to assume these 3rd party libs will be installed separately from the plugin? Or would they need to be included in packaging ref such as `setup_requirements.txt` or `setup.py`?

File: f5_openstack_agent/utils/cert_parser.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    :param intermediates: PEM or PKCS7 encoded intermediate certificates
+    :returns: A list of strings where each string represents an
+              X509 pem block surrounded by BEGIN CERTIFICATE,
+              END CERTIFICATE block tags
+    """
+    if X509_BEG in intermediates:
+        for x509Pem in _split_x509s(intermediates):
+            yield _prepare_x509_cert(_get_x509_from_pem_bytes(x509Pem))
+    else:
+        for x509Pem in _parse_pkcs7_bundle(intermediates):
```
Comment (kaustriaf5) [inline f5_openstack_agent/utils/cert_parser.py]: `x509Pem` is actually `x509Der`

File: f5_openstack_agent/utils/cert_parser.py (hunk 2 above)
Comment (zhang-shengping) [reply f5_openstack_agent/utils/cert_parser.py]: I am not sure that we need to add all test cases in the future code for this repo?

When I run the tox unit test, I found some of the previous tests (test_l2_service.py, test_network_helper_delete_fdb_entries) ran wrong, and I am sure it's not caused by my code changes. 

I will fix it as a bug if no one to fix it.

File: f5_openstack_agent/utils/cert_parser.py (hunk 3 above)
Comment (zhang-shengping) [reply f5_openstack_agent/utils/cert_parser.py]: I checked the modules, they are in Openstack global requirements https://github.com/openstack/requirements. 
I think it is better to write in the setup_requirements.txt

Comment (zhang-shengping) [reply f5_openstack_agent/utils/cert_parser.py]: It seems the test won pass when I set the setup_requirements.txt and requirements.unittest.txt file.

I pushed three times.
The first time, I add below content to requirements.unittest.txt and  setup_requirements.txt:
       six>=1.9.0 # MIT
       pyasn1>=0.1.9,<=0.4.3
       pyasn1-modules>=0.0.8,<=0.2.2
       cryptography>=1.2.3,<=2.3

error:
`dpkg-gencontrol: warning: can't parse dependency python-pyasn1 (<=0.4.3
dpkg-gencontrol: error: error occurred while parsing Depends field: , python (>= 2.7), python (<< 2.8), python:any (>= 2.7.1-0ubuntu2), python:any, python-f5-sdk, python-f5-sdk (<< 3.0.12), python-f5-sdk (>= 3.0.11), python-six (>=1.9.0), python-pyasn1 (<=0.4.3, >=0.1.9), python-pyasn1-modules (<=0.2.2, >=0.0.8), python-cryptography (<=2.3, >=1.2.3)`


The second time, I add below content to requirements.unittest.txt and  setup_requirements.txt:
       six>=1.9.0 # MIT
       pyasn1<=0.4.3
       pyasn1-modules<=0.2.2
       cryptography<=2.3

error:
`Can't get package dependencies for f5-openstack-agent-10.2.0-1.el7.noarch.rpm
 -- rpm -ivh /tmp/*.rpm
Agent install failed
The command "f5-openstack-agent-dist/scripts/test_install.sh "redhat" "7" ${PKG_RELEASE_EL7}" exited with 1.`


The third time (this time), I add below content to requirements.unittest.txt and  setup_requirements.txt:
       six>=1.9.0 # MIT
       pyasn1>=0.1.9
       pyasn1-modules>=0.0.8
       cryptography>=1.2.3

error:
`Failed to copy f5-openstack-agent package
Getting dependencies for f5-openstack-agent-10.2.0-1.el7.noarch.rpm.
 -- rpm -qRp f5-openstack-agent-dist/rpms/build/f5-openstack-agent-10.2.0-1.el7.noarch.rpm
Can't get package dependencies for f5-openstack-agent-10.2.0-1.el7.noarch.rpm
 -- rpm -ivh /tmp/*.rpm
Agent install failed
The command "f5-openstack-agent-dist/scripts/test_install.sh "redhat" "7" ${PKG_RELEASE_EL7}" exited with 1.`

`Installing: /tmp/python-f5-sdk-rest_3.0.11_1404_all.deb
SDK Succeeded in install test
Installing: /tmp/python-f5-openstack-agent_10.2.0-1_1404_all.deb
SDK install failed (1)
The command "f5-openstack-agent-dist/scripts/test_install.sh "ubuntu" "14.04" ${PKG_RELEASE_1404}" exited with 1.`

I am not very sure why this happened. Do you have any idea to check the process in detail or why did it happen?

Comment (zhang-shengping) [reply f5_openstack_agent/utils/cert_parser.py]: @kaustriaf5

Comment (zhang-shengping) [reply f5_openstack_agent/utils/cert_parser.py]: The problem happens when it install the packages on Redhat using rpm and Ubuntu using dpkg.

File: f5-openstack-agent-dist/Docker/ubuntu/install_test/fetch_and_install_deps.py [hunk 5]
```diff
@@ -24,6 +24,8 @@ def runCommand(cmd):
                              stdout=subprocess.PIPE,
                              stderr=subprocess.PIPE)
         (output) = p.communicate()[0]
+        #(pzhang) wait the process finish
```
Comment (zhaoqin-github) [inline f5-openstack-agent-dist/Docker/ubuntu/install_test/fetch_and_install_deps.py]: Do we really need this?

File: setup.cfg [hunk 6]
```diff
@@ -0,0 +1,5 @@
+[bdist_rpm]
```
Comment (zhaoqin-github) [inline setup.cfg]: remove this file

File: f5_openstack_agent/lbaasv2/drivers/bigip/listener_service.py [hunk 7]
```diff
@@ -177,20 +177,28 @@ def _create_ssl_profile(
...
+            chain = '\n'.join(list(intermediates))
+
         name = self.cert_manager.get_name(container_ref,
                                           self.service_adapter.prefix)
 
         try:
             # upload cert/key and create SSL profile
             ssl_profile.SSLProfileHelper.create_client_ssl_profile(
-                bigip, name, cert, key, sni_default=sni_default,
+                bigip, name, cert, key, intermediates=chain,
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/listener_service.py]: If possible, make parameter order to be same as their declaration order.

**Comment by zhaoqin-github** [comment]
@zhang-shengping you'd better use a new commit id to apply new changes, after you "merge master", instead of "--amend" the merge commit. I suggest you to reset your 2nd commit id and commit your changes again.

**Pull request merged**